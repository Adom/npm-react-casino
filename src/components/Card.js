export default ({ suit, face, width, height, className }) => {
  const defaultClasses = ['playing-card'];
  let style = {};
  if (width)
    style.width = width;
  if (height)
    style.height = height;
  if (!width && !height)
    style.width = 150;
  suit = String(suit).toUpperCase();
  face = String(face).toUpperCase();
  switch(suit) {
    case 'C':
    case 'D':
    case 'H':
    case 'S':
      break;
    default: 
      suit = '';
  }
  switch(suit.length && face) {
    case 'A':
    case 'K':
    case 'Q':
    case 'J':
    case 'T':
    case '9':
    case '8':
    case '7':
    case '6':
    case '5':
    case '4':
    case '3':
    case '2':
      break;
    default: 
      face = 'BACK';
  }
  return (
    <span className={ Array.isArray(className) ? [...className, ...defaultClasses].join(' ') : className || defaultClasses.join(' ') }>
      <img src={ require(`./components/png/cards/${face}${suit}.png`) } alt={`${face}${suit}`} style={style} />
    </span>
  );
};