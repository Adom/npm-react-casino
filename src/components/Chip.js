export default ({ color, size=52 }) => {
  return (
    <div>Chip</div>
  );
};