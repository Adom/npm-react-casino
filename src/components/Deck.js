export default ({ color, size=52 }) => {
  return (
    <div>Deck</div>
  );
};