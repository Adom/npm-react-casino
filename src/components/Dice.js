export default ({ color, size=52 }) => {
  return (
    <div>Dice</div>
  );
};