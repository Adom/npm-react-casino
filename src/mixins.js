import React from 'react';

export const cardValue = (face) => {
  switch(face) {
    case 'A': return 14;
    case 'K': return 13;
    case 'Q': return 12;
    case 'J': return 11;
    case '10': return 10;
    case '9': return 9;
    default: return 0;
  }
}

export const reducedClassNames = (defaultClasses=[], className=[]) => Array.isArray(className) ? [...className, ...defaultClasses].join(' ') : className || defaultClasses.join(' ');

export const isBower = (card,trump,bowers={S: 'C', C: 'S', H: 'D', D: 'H'}) => trump && card.suit === bowers[trump] && card.face === 'J';

export const isPlayable = (card,hand,suit,trump) => {
  const trumpBower = isBower(card,trump);
  const sameSuit = card.suit === suit;
  const noCardLed = !suit;
  let haveNoneOfSuit = true;
  hand.map(c => {
    if (c.suit === suit && !isBower(c,trump)) {
      haveNoneOfSuit = false;
    }
    return c;
  });
  const isPlayable = noCardLed || sameSuit || (trumpBower && suit === trump) || haveNoneOfSuit;
  return isPlayable;
};


export const cardClassNames = (card,hand,suit,trump,swaps=[]) => {
  let classNames = [];
  if (!isPlayable(card,hand,suit,trump)) {
    classNames.push('dim');
  } else {
    classNames.push('playable');
  }
  // check swaps
  if (swaps && swaps.length === 1 && swap) {
    const thisCardSwaps = swaps.filter(swap => swap.card.suit === card.suit && swap.card.face === card.face);
    if (thisCardSwaps.length) {
      classNames.push('swapped');
    }
  }
  return classNames.join(' ');
};

export const sortHand = (cards, trump=false) => {
  // console.log('[&] Sorting cards:', cards);
  let bySuit = {};
  for(let card of cards) {
    if (!bySuit[card.suit]) {
      bySuit[card.suit] = [];
    }
    bySuit[card.suit].push(card);
  }
  for(let suit in bySuit) {
    bySuit[suit].sort((a,b) => cardValue(b.face) - cardValue(a.face));
  }
  let sorted = [];
  if (bySuit['C']) {
    sorted = [
      ...sorted,
      ...bySuit['C'],
    ];
  }
  if (bySuit['S']) {
    sorted = [
      ...sorted,
      ...bySuit['S'],
    ];
  }
  if (bySuit['D']) {
    sorted = [
      ...sorted,
      ...bySuit['D'],
    ];
  }
  if (bySuit['H']) {
    sorted = [
      ...sorted,
      ...bySuit['H'],
    ];
  }
  // console.log('[&] Sorted cards:', sorted);
  return sorted;
};