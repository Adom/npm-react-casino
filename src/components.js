import React from 'react';
import { cardClassNames, reducedClassNames } from './mixins';
import cardsCss from './assets/css/cards.css';
import handsCss from './assets/css/hands.css';
/******************************************************************
 * Styles
 ******************************************************************/
export const CardStyles = () => (
  <style dangerouslySetInnerHTML={{__html: cardsCss}} />
);
export const HandStyles = () => (
  <style dangerouslySetInnerHTML={{__html: handsCss}} />
);
/******************************************************************
 * Deck
 ******************************************************************/
export const Deck = ({ color, size=52 }) => {
  return (
    <div>Deck</div>
  );
};
/******************************************************************
 * Hand
 ******************************************************************/
export const Hand = ({ cards=[], trump=false, follow=false, onClick=()=>{}, onHover=()=>{} }) => {
  const defaultClasses = ['playing-card'];
  return (
    <ul className={ reducedClassNames(className, defaultClasses) }>
      {
        cards.map((card,i) => {
          const classes = cardClassNames(card,cards,follow,trump);
          return (
            <li key={i} className={classes}>
              <Card suit={card.suit} face={card.face} playable={classes.includes('playable')} onClick={onClick} onHover={onHover} />
            </li>
          );
        })
      }
    </ul>
  );
};
/******************************************************************
 * Card
 ******************************************************************/
export const Card = ({ suit, face, width, height, className, onClick=()=>{}, onHover=()=>{} }) => {
  const defaultClasses = ['playing-card'];
  let style = {};
  if (width)
    style.width = width;
  if (height)
    style.height = height;
  if (!width && !height)
    style.width = 150;
  suit = String(suit).toUpperCase();
  face = String(face).toUpperCase();
  switch(suit) {
    case 'C':
    case 'D':
    case 'H':
    case 'S':
      break;
    default: 
      suit = '';
  }
  switch(suit.length && face) {
    case 'A':
    case 'K':
    case 'Q':
    case 'J':
    case 'T':
    case '9':
    case '8':
    case '7':
    case '6':
    case '5':
    case '4':
    case '3':
    case '2':
      break;
    default: 
      face = 'BACK';
  }
  const click = (e,card) => {
    onClick(e,card);
  };
  const hover = (e,card) => {
    onHover(e,card);
  };
  return (
    <span onClick={e => click(e,{face,suit})} onHover={e => hover(e,{face,suit})}
      className={ reducedClassNames(className,defaultClasses) }>
      <img src={ require(`./assets/png/cards/${face}${suit}.png`) } alt={`${face}${suit}`} style={style} />
    </span>
  );
};
/******************************************************************
 * Chip
 ******************************************************************/
export const Chip = ({ color, value }) => {
  return (
    <div>Chip</div>
  );
};
/******************************************************************
 * Dice
 ******************************************************************/
export const Dice = ({ color, value }) => {
  return (
    <div>Dice</div>
  );
};