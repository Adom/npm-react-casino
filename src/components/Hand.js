import { cardClassNames } from '..//mixins';
export default ({ cards=[], trump=false, playable=false, fan=10, space=5 }) => {
  return (
    <ul className="hand">
      {
        cards.map((card,i) => (
          <li key={i} className={cardClassNames(card)}>
            <Card suit={card.suit} face={card.face} playable={playable} />
          </li>
        ))
      }
    </ul>
  );
};