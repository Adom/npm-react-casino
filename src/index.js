import * as Mixins from './mixins';
import * as Components from './components';

export default {
  ...Mixins,
  ...Components
};
