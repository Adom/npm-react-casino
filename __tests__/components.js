import React from 'react';
import renderer from 'react-test-renderer';
import { Card } from '../src/components';
test('Card will render', () => {
  const component = renderer.create(
    <Card suit="S" face="A" />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseEnter();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();

  // // manually trigger the callback
  // tree.props.onMouseLeave();
  // // re-rendering
  // tree = component.toJSON();
  // expect(tree).toMatchSnapshot();
});