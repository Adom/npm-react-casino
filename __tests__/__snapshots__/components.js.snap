// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card will render 1`] = `
<span
  className="playing-card"
  onClick={[Function]}
  onHover={[Function]}
>
  <img
    alt="AS"
    src={
      Object {
        "process": [Function],
      }
    }
    style={
      Object {
        "width": 150,
      }
    }
  />
</span>
`;
